implemented native SASL/JAAS support for plaintext authentication

schulzp opened the pull request:
fixes #8168

cescoffier commented on extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/KafkaProcessor.java:
```diff
@@ -68,6 +76,9 @@
             FloatDeserializer.class,
+
+            //login modules
+            LoginModule.class
```
This list was about serde, LoginModule might go in another list (it's actually not build-in as it comes from `javax.security.auth.spi.LoginModule`)

schulzp replied:
Would a separate array/list suffice?

cescoffier commented on extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/KafkaProcessor.java:
```diff
@@ -109,6 +120,12 @@ public void build(CombinedIndexBuildItem indexBuildItem, BuildProducer<Reflectiv
         reflectiveClass.produce(new ReflectiveClassBuildItem(true, false, "sun.misc.Cleaner"));
+
+        // enable native JAAS/SASL PLAIN support
+        reflectiveClass
```
I would add the LoginModule directly here.

schulzp replied:
Actually, I need all the sub-classes, for example (but not limited to):
* PlainLoginModule (org.apache.kafka.common.security.plain)

The list of sub-classes also contains:
* ExternalCertificateLoginModule (org.apache.activemq.artemis.spi.core.security.jaas)
* Krb5LoginModule (org.apache.activemq.artemis.spi.core.security.jaas)
* PropertiesLoginModule (org.apache.activemq.artemis.spi.core.security.jaas)
* DigestLoginModule (org.apache.zookeeper.server.auth)
* LDAPLoginModule (org.apache.activemq.artemis.spi.core.security.jaas)
* ScramLoginModule (org.apache.kafka.common.security.scram)
* NTLoginModule (com.sun.security.auth.module) 
* GuestLoginModule (org.apache.activemq.artemis.spi.core.security.jaas)
* InVMLoginModule (org.apache.activemq.artemis.spi.core.security.jaas)
* FileLoginModule (com.sun.jmx.remote.security)
* JndiLoginModule (com.sun.security.auth.module)
* AbstractKeycloakLoginModule (org.keycloak.adapters.jaas)
* LdapLoginModule (com.sun.security.auth.module)
* OAuthBearerLoginModule (org.apache.kafka.common.security.oauthbearer)
* UnixLoginModule (com.sun.security.auth.module)
* Krb5LoginModule (com.sun.security.auth.module)
* CertificateLoginModule (org.apache.activemq.artemis.spi.core.security.jaas)
* KeyStoreLoginModule (com.sun.security.auth.module)

schulzp replied:
Relocated native-SASL/JAAS code into separate build step.

cescoffier commented on extensions/kafka-client/runtime/src/main/java/io/quarkus/kafka/client/health/KafkaHealthCheck.java:
```diff
@@ -3,36 +3,39 @@
-        conf.put(AdminClientConfig.BOOTSTRAP_SERVERS_CONFIG, bootstrapServers);
+    @Inject
+    public KafkaHealthCheck(KafkaRuntimeConfig runtimeConfig) {
+        Map<String, Object> conf = new HashMap<>(runtimeConfig.getProperties());
```
How is this related to the issue? I may miss a link.

schulzp replied:
the previous `@PostConstruct` ran into NPEs for the `@Inject`ed `KafkaRuntimeConfig`. Using constructor injection it worked.

cescoffier replied:
Why not using the static access to the configuration? That would remove the bean.

cescoffier commented on extensions/kafka-client/runtime/src/main/java/io/quarkus/kafka/client/runtime/KafkaRuntimeConfigProducer.java:
```diff
@@ -0,0 +1,43 @@
+@Dependent
+public class KafkaRuntimeConfigProducer {
+
+    private String configPrefix = "kafka";
```
If I understand correctly you copy the "kafka" configuration and configure the health check admin client with them.
Is it really related to this PR?

schulzp replied:
I'd say so. The admin client needs the same security configuration as all consumers/producers. Otherwise authentication will fail once the broker is set up to expect it.

cescoffier replied:
Yes and no. You have applications using multiple brokers, and some parts of Quarkus use a different configuration format.

cescoffier commented on extensions/kafka-client/runtime/src/main/java/io/quarkus/kafka/client/runtime/KafkaRuntimeConfigProducer.java:
```diff
@@ -0,0 +1,43 @@
+        StreamSupport
+                .stream(config.getPropertyNames().spliterator(), false)
+                .map(String::toLowerCase)
+                .filter(name -> name.startsWith(configPrefix))
```
It can be confusing as the Kafka Client is not produced by default. Typically, it would not cover the configuration coming from reactive messaging. Existing applications create their properties directly (following the Kafka documentation).

schulzp replied:
Okay, in that case I could remove the KafkaRuntimeConfiguration again and would only have to make sure all the clients in integration-tests/kafa get the right set of properties. I just found it to be very cumbersome to inject them one by one, especially when extending the test case to more than plain authentication.

cescoffier commented on integration-tests/kafka/pom.xml:
```diff
@@ -114,13 +135,31 @@
                                 <configuration>
                                     <systemProperties>
-                                        <native.image.path>${project.build.directory}/${project.build.finalName}-runner</native.image.path>
+                                        <kafka.security.protocol>SASL_PLAINTEXT</kafka.security.protocol>
```
Why is this passed here and not in the Map produced by the Kafka Test Resource?

schulzp replied:
This *instructs* the test resource to set up the broker to request authentication and to pass the required kafka-client properties on to quarkus.

schulzp replied:
Implemented dedicated integration test, leaving the existing ones untouched.

cescoffier commented on integration-tests/kafka/pom.xml:
```diff
@@ -96,6 +96,27 @@
+                        </goals>
+                        <configuration>
+                            <systemProperties>
+                                <kafka.security.protocol>SASL_PLAINTEXT
```
Why are these properties configured here and not in the Kafka Test Resource?

schulzp replied:
This instructs the test resource to set up the broker to request authentication and to pass the required kafka-client properties on to quarkus.

schulzp replied:
Implemented dedicated integration test, leaving the existing ones untouched.

cescoffier commented on extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/KafkaProcessor.java:
```diff
@@ -130,4 +147,13 @@ HealthBuildItem addHealthCheck(KafkaBuildTimeConfig buildTimeConfig) {
+
+    @BuildStep
+    AdditionalBeanBuildItem addRuntimeConfig() {
+        final Builder builder = AdditionalBeanBuildItem.builder()
```
I believe we should be able to avoid this new bean. 
It seems to be convenient for the tests, but the Kafka Test Resource can configure properties (that's the Map returned by the `start` method)

cescoffier replied:
If we go with this, we would need to adapt the Reactive Messaging Kafka connector to use the same config. This would be a breaking change.

schulzp committed:
#8817 - change names for clarity (SSL != SASL)

schulzp commented:
@cescoffier, thank you for your review! I fixed a few things:
1. I followed your SASL/SSL example regarding the integration tests. You already named your tests SASL[...] despite only enabling SSL so I allowed myself to rename them into SSL[...] and used SASL for actual SASL/JAAS authentication.
2. I moved all native-SASL/JAAS related code into a separate `@BuildStep`.
3. I still don't know how to make `KafkaHealthCheck` or more specific its `AdminClient` aware of the properties like required to connect to a SASL/JAAS-enabled broker, for example, `sasl.jaas.config`, `sasl.mechanism`, and `security.protocol`.

cescoffier reviewed (changes requested):
Almost there! Looks mostly good to me.

Still wondering about the new bean introduced to carry the config. I'm not against it, it just needs to be done carefully because the RM Kafka connector uses a different format and, because of project dependencies (Quarkus depends on the connector not the opposite), would not be able to depend on a Quarkus bean. My concern is that with such beans, the config would be quite different between the health checks and the connector. 

For AMQP we introduced the same kind of bean (but in the connector and with name support to handle the multi-broker case). 

Basically, what we would need is a bean producing a Map<String, ?>. If we use the `@Named` qualifier and set it to "default-kafka", we can imagine having the health check retrieving and using this bean and the RM connector retrieving and using this bean too. Would that work?

schulzp commented:
Hi @cescoffier,

I've experimented a bit with code generation to cater for multiple (named) brokers but I keep running into alleged dead ends. Here is what I did:
* Create a `KafkaBrokerBuildTimeConfig` and add `defaultBroker` and `namedBrokers` to `KafkaBuildTimeConfig`
* Generate a `KafkaRuntimeConfigProducer extends AbstractKafkaRuntimeConfigProducer` with `@Produces @Named("kafka-config-{broker}") Map<String, ?> produce_{broker}() { ... }` per (named) broker.
* Generate a `KafkaHealthCheck_{broker}` with `@Inject @Named("kafka-config-{broker}") Map<String, ?> config`

Problems I ran into:
* Despite specifying a super class for the generated classes they cannot be cast to it (ClassCastException) when the `@Recorder` should call `AbstractKafkaRuntimeConfigProducer#setBuildTimeConfig`
* The `KafkaBrokerBuildTimeConfig` currently has only one `@ConfigProperty(name = "security.protocol")` to determine at build-time if I need to load all the SASL/JAAS-related classes. In a scenario where no broker uses a non-default `security.protocol` the list of named broker configurations is empty at build-time and to config/health-check gets generated. When I later specify `kafka.broker_a.boostrap.servers=...` and `kakfa.broker_b.boostrap.servers` at runtime that fails because there is no `KafkaRuntimeConfigProducer` which picks up the config for `broker_a` or `broker_b`.
* I only want to generate a health-check if the health feature is enabled. If I pass `ClassCreator#getClassName()` to `HealthBuildItem` it fails at runtime since the class-loader cannot find the generated class.

Is this the right approach? Any help would be appreciated.

cescoffier commented:
Did you push the code? Maybe we should start with just a Map exposed using `@Named("default-kafka-broker")` and we revisit later. Note that this map should be exposed regardless of health as the RM connector would consume it.

schulzp commented:
@cescoffier, pushed it now with runtime config being exposed as `@Named("default-kafka-broker") Map<String, Object>`. Hopefully that works for you.

cescoffier commented:
Yes, that will work for me! Thanks! I will change the Kafka connector to consume this map.

schulzp commented:
Hi @cescoffier, the CI pipelines passed. Do you still require any changes in the PR?

cescoffier commented on integration-tests/kafka/pom.xml:
```diff
@@ -196,6 +196,9 @@
                         <configuration>
                             <skip>false</skip>
+                            <systemProperties>
+                                <quarkus.debug.generated-sources-dir>${project.build.directory}/generated-sources</quarkus.debug.generated-sources-dir>
```
Any reason for this?

schulzp replied:
Good catch, removed it.

cescoffier reviewed (changes requested):
One small comment.

Could you also rebase on latest master?

schulzp committed:
#8817 - runtime configuration as map

schulzp commented:
@cescoffier, done.

schulzp commented:
Hi @cescoffier, rebased again, resolved conflicting change 0dc12512fa6b26aefd0fd85461a40a7377e74034 in `KafkaSASLTestResource`/`KafkaSSLTestResource`

cescoffier commented:
@schulzp Waiting for CI, will merge once the CI says `ok`.

cescoffier commented:
Thanks @schulzp !
